### ISSUE TITLE
Tell user when job hangs due to lack of resources

### ISSUE THREAD

**Issue by glennhickey** [DataBiosphere/toil#3016]
When Toil can't schedule a job due to lack of resources, it just hangs indefinitely.  If the user is not running `--logDebug`, then they never get told why their job has stopped.  

This comes up frequently in Cactus, ex:
https://github.com/ComparativeGenomicsToolkit/cactus/issues/150
https://github.com/ComparativeGenomicsToolkit/cactus/issues/182

Can Toil please make a clear warning for these situations at a higher log level?  

In general it's hard to detect a deadlock, but I think on `singleMachine` if it's in a state where no jobs are running, and no jobs can be run, then exiting with a clear error would be much better than the current behaviour.

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-530)
┆Issue Number: TOIL-530

**Comment by Robin-Rounthwaite** [comment]
I also encountered this issue. Took me some time to figure out what was going wrong.

**Comment by adamnovak** [comment]
The `SingleMachineBatchSystem` already detects if jobs are too big to ever run. It does cores and memory and the `AbstractBatchSystem` also does disk:

https://github.com/DataBiosphere/toil/blob/d4c7862832ba1faa2ef89336e8ed251919cefc2b/src/toil/batchSystems/singleMachine.py#L485-L494

I can revise/simplify this code so we don't have two copies of it, but at least in https://github.com/ComparativeGenomicsToolkit/cactus/issues/150 it seems to already be working.

I don't think either of those reported issues are states in which no jobs are running. It looks like it is managing to start at least a service job, and then the service job deadlock detector is not detecting the deadlock, or not detecting it promptly.

That first issue has a log that looks like:

```
Limiting maxCores to CPU count of system (4).
Limiting maxMemory to physically available memory (8274743296).
Limiting maxDisk to physically available disk (2799091712).
...
Issued job 'KtServerService' kind-KtServerService/instancee7dxjqzv with job batch system ID: 151 and cores: 0, disk: 2.0 G, and memory: 2.3 G
INFO:toil.worker:Redirecting logging to /tmp/toil-358a4b86-1e2c-415a-a5c4-0969e1d01e55-3a82b8c940344fe39ed4b367a7e64d29/tmp7ui031wr/worker_log.txt
Issued job 'CactusSetupPhase' kind-CactusSetupPhase/instancepvtt5lwa with job batch system ID: 152 and cores: 1, disk: 2.0 G, and memory: 3.3 G
Reissued any over long jobs
Reissued any over long jobs
INFO:toil.worker:Redirecting logging to /tmp/toil-358a4b86-1e2c-415a-a5c4-0969e1d01e55-3a82b8c940344fe39ed4b367a7e64d29/tmpu1o8s6m1/worker_log.txt
Traceback (most recent call last):
...
```

The system has 2.7 GB of disk space to use, and it's not until we get to the part of the workflow where we try to run a 2 GB service and a 2 GB service user that we have a deadlock. But after that point it appears to have taken us two ticks of the reissue loop (which [ticks every hour](https://github.com/DataBiosphere/toil/blob/727b7e1c30938c575717b46f0bc9f88758e3b6e9/src/toil/common.py#L119)) to actually detect the deadlock, even though we're supposed to be looking every minute or so.

But the fragments of the debug log we got in https://github.com/ComparativeGenomicsToolkit/cactus/issues/150#issuecomment-589935706 make it look like it was actually running a job the whole time, and it finally finished after 2 hours, prompting the deadlock to happen. The debug log isn't really consistent with the original log as far as I can tell. I will have to see if I can reproduce this myself. @glennhickey do you have an idea of how long the different phases of that test workflow are supposed to take? I think I ran it once and it didn't take 2 hours.

The [second issue](https://github.com/ComparativeGenomicsToolkit/cactus/issues/182) has a log that starts out:

```
Limiting maxCores to CPU count of system (8).
Limiting maxMemory to physically available memory (8057978880).
Limiting maxDisk to physically available disk (16912220160).
```

We [shouldn't see that](https://github.com/DataBiosphere/toil/commit/0283b1c0a9b01ed31ca1e440b6ae3323393769e0#diff-9c70356a721a1c2fb810f9234557ceb9L49-L51) if @jmonlong really passed `--maxCores 1`, as far as I can tell. And that's 8 cores, 8 GB memory, and 16 GB disk; isn't that enough for the workflow?

I'm going to have to try and replicate these myself; as far as I can tell we already should be catching and bailing on these sorts of deadlocks.

**Comment by glennhickey** [comment]
The test workflow in #150 (with `--root mr`) should take under an hour on pretty much any system.  The fact that it's that long is annoying but is another issue https://github.com/ComparativeGenomicsToolkit/cactus/issues/219).

Anyway, it does seem to be detecting these cases and (eventually) exiting with deadlock exceptions.  I agree that the hang in the first log of #182 looks unrelated to resources.  Robin's log (https://docs.google.com/document/d/1M3WCOEFaE1rB60EYGpem3Kva2vzPTyMbpMXbMfZ7AVA/edit) is similar to #150 but he was blocked on disk.  

So the main toil issue then as far as I see isn't so much detection of deadlocks (sorry!), but reporting their causes.  Right now, you only see that it's blocked on whatever resource by poking through `--logDebug` output, which isn't something we can expect of any normal user.

**Comment by adamnovak** [comment]
OK, I have a lead on *something*. When I do:

```
mkdir -p ~/mnt/tmpfs
sudo mount -t tmpfs -o size=2200000000 tmpfs ~/mnt/tmpfs
TMPDIR=$HOME/mnt/tmpfs cactus --workDir=$HOME/mnt/tmpfs ./tree examples/evolverMammals.txt examples/evolverMammals.hal --root mr
```

I see it getting into what appears to be a deadlock that the deadlock detector doesn't detect:

```
Job ended: 'CactusSetupCheckpoint' kind-CactusSetupCheckpoint/instance-lxh2es7v
Issued job 'StartPrimaryDB' kind-StartPrimaryDB/instance-8etdxtig with job batch system ID: 162 and cores: 1, disk: 2.0 G, and memory: 3.3 G
INFO:toil.worker:Redirecting logging to /z/home/anovak/mnt/tmpfs/node-e51ea818-4587-419d-b346-78aaa5676ac5-11c2d78d2bcc48f28d74bd1f7dc0a45f/tmps2b3vrcr/worker_log.txt
Job ended: 'StartPrimaryDB' kind-StartPrimaryDB/instance-8etdxtig
Issued job 'KtServerService' kind-KtServerService/instance-fefmp7lb with job batch system ID: 163 and cores: 0, disk: 2.0 G, and memory: 2.3 G
INFO:toil.worker:Redirecting logging to /z/home/anovak/mnt/tmpfs/node-e51ea818-4587-419d-b346-78aaa5676ac5-11c2d78d2bcc48f28d74bd1f7dc0a45f/tmpeoyz9mj7/worker_log.txt
Issued job 'CactusSetupPhase' kind-CactusSetupPhase/instance-0zxc90us with job batch system ID: 164 and cores: 1, disk: 2.0 G, and memory: 3.3 G
```

It sits there for several minutes at least at this stage. I think the database server comes up but nothing can run against it.

I shall investigate.

**Comment by adamnovak** [comment]
Turns out this is Cactus's own dang fault. It forces Toil to wait for at least an hour between polling for deadlocks, no matter what the user says on the command line about `--deadlockWait`:

https://github.com/ComparativeGenomicsToolkit/cactus/blob/3ec10eff0e4da3368a5e100b0df1f91cd56370ad/src/cactus/progressive/cactus_progressive.py#L391-L396

Since Toil polls once almost immediately on startup (since it hasn't done it yet), and since detecting a deadlock requires seeing a potential deadlock and then seeing it still stuck later, it is going to take a minimum of 2 hours for Cactus to realize it has filled everything up with service jobs, no matter the batch system it is running on.

(It also appears to be forcing caching off, so if we ever thought we tested Cactus with caching I suspect we didn't.)

I can revise Toil so it starts announcing some warnings earlier, and make `--deadlockWait` just how long it has to see a deadlock for to actually kill the workflow.

**Comment by jmonlong** [comment]
I don't know if it was related to the problem I noticed on my small laptop but if it helps I can try to run a fix on this same machine at some point.
 
BTW here was the more comprehensive log I got when running it with the appropriate number of cores but still getting "stuck":
[hsvlr-100bp_2_ins-cactus.log.txt](https://github.com/DataBiosphere/toil/files/4518459/hsvlr-100bp_2_ins-cactus.log.txt)

I had run with the data from [my cactus issue](https://github.com/ComparativeGenomicsToolkit/cactus/issues/182) : `docker run -it -v `pwd`:/app -u `id -u $USER` -w /app jmonlong-cactus cactus jobStore_hsvlr-100bp_2_ins test.txt test.hal --binariesMode local --maxCores 3 --logFile hsvlr-100bp_2_ins-cactus.log --realTimeLogging --logInfo`

(I'm not blocked at all by this issue anymore because I moved to a bigger machine so if what you're discussing is not related please disregard what I just said).

**Comment by diekhans** [comment]
Well that sure is tacky of Cactus.  I will file this as a bug in Cactus.

If early the warnings are false, they could create a sense of
alarm; so if you do change the behavior, maybe word it so it
inspires bliss rather than alarm.
